File: 0301-0400/0377.combination-sum-iv.cpp
#include "common.h"

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
      vector<unsigned int> dp(target + 1);
      dp[0] = 1;
      for (int i = 1; i <= target; ++i) {
        for (int x : nums) {
          if (i >= x)
            dp[i] += dp[i - x];
        }
      }
      return dp[target];
    }
};

int main() {
  vector<int> nums = {1, 2, 3};
  int target = 4;
  cout << Solution().combinationSum4(nums, target) << endl;
  return 0;
}

File: 0301-0400/0377-combination-sum-iv.py
from typing import List

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        '''
        注意, 这里问的是排列数
        '''
        dp = [0] * (target + 1)
        dp[0] = 1
        for i in range(1, target + 1):
            for x in nums:
                if i >= x:
                    dp[i] += dp[i - x]
        return dp[target]

print(Solution().combinationSum4([1, 2, 3], 4))
